nomad/eval_broker_test.go

```diff
@@ -1506,3 +1383,40 @@ func TestEvalBroker_NamespacedJobs(t *testing.T) {
 	require.Equal(1, len(b.blocked))
 
 }
+
+func TestEvalBroker_BlockedEvals_MarkForCancel(t *testing.T) {
```

Comment:
I wonder if we could add an integration test to avoid creating unrealistically ideal evals and indexes in tests.

Since we can manipulate NumSchedulers on a live server it should be possible to start a TestServer or TestAgent with num_schedulers=0 and create some interesting pending evals by:

1. Registering a system job (creates an initial eval)
2. Register a node (creates a second eval for the system job)
3. Set num_schedulers=1
4. ...wait...
5. Assert the initial eval was `complete`'d successfully and the second eval was `canceled`

I think that should assert that your optimization works in somewhat realistic conditions. Obviously we aren't concerned with folks actually flipping num_schedulers, but I think that's a reasonable bit of unrealistic interaction to let us get a queue of evals for a single job in such a way that is currently *not* optimized but would *should be optimized by this PR.*

Reply:
> I wonder if we could add an integration test to avoid creating unrealistically ideal evals and indexes in tests.

Yeah this updated `TestEvalBroker_Serialize_DuplicateJobID` test is totally broken b/c it's assuming that a change to the Eval's modify index is at all visible to the broker, which it is not. I'm not sure how I got it in my head that we were getting an update Eval. 😊

Reply:
Done in `TestEvalBroker_IntegrationTest`. I've got it so that it flaps the node status and then asserts that we only get 2 evals processed (the first one and then the one from the top of the `blocked` queued once that one is ack'd), with the rest being canceled.